Realm/RLMObjectStore.mm

```diff
@@ -345,26 +343,31 @@ static inline void RLMVerifyInWriteTransaction(RLMRealm *realm) {
     RLMCheckThread(realm);
 }
 
-void RLMInitializeSwiftListAccessor(RLMObjectBase *object) {
-    // switch List<> properties to linkviews from standalone arrays
-    static Class s_swiftObjectClass = NSClassFromString(@"RealmSwift.Object");
-    if (!object || !object->_row || ![object isKindOfClass:s_swiftObjectClass]) {
+void RLMInitializeSwiftListAccessor(__unsafe_unretained RLMObjectBase *const object) {
+    if (!object || !object->_row || !object->_objectSchema.isSwiftClass) {
         return;
     }
 
+    static Class s_swiftObjectClass = NSClassFromString(@"RealmSwift.Object");
+    if (![object isKindOfClass:s_swiftObjectClass]) {
+        return; // Is a Swift class using the obj-c API
+    }
+
     for (RLMProperty *prop in object->_objectSchema.properties) {
         if (prop.type == RLMPropertyTypeArray) {
-            [RLMObjectUtilClass(YES) initializeListProperty:object property:prop array:[RLMArrayLinkView arrayWithObjectClassName:prop.objectClassName
-                                                                                                                             view:object->_row.get_linklist(prop.column)
-                                                                                                                            realm:object->_realm]];
+            RLMArray *array = [RLMArrayLinkView arrayWithObjectClassName:prop.objectClassName
+                                                                    view:object->_row.get_linklist(prop.column)
+                                                                   realm:object->_realm];
+            [RLMObjectUtilClass(YES) initializeListProperty:object property:prop array:array];
         }
     }
 }
 
-} // extern "C" {
-
 template<typename F>
-static inline NSUInteger RLMCreateOrGetRowForObject(RLMObjectSchema *schema, F primaryValueGetter, RLMCreationOptions options, bool &created) {
+static inline NSUInteger RLMCreateOrGetRowForObject(__unsafe_unretained RLMObjectSchema *const schema,
+                                                    F&& primaryValueGetter,
```

Comment:
Makes it take a reference to the temporary lambda rather than copying it, since the retain/release on the captured obj-c object aren't optimized out.